concurrency/DCMetricsUtils.java

```diff
@@ -0,0 +1,481 @@
+        dcmConfig = newConfig;
+        serviceName = dcmConfig.getServiceName();
+        methodName = dcmConfig.getMethodName();
+        teamName = dcmConfig.getTeamName();
+        siteVariant = dcmConfig.getSiteVariant();
+        mFactory = AndroidMetricsFactoryImpl.getInstance(context);
+        AndroidMetricsFactoryImpl.setOAuthHelper(context, dcmConfig.getOAuthHelper());
+        AndroidMetricsFactoryImpl.setDeviceType(context, dcmConfig.getDeviceType());
+        AndroidMetricsFactoryImpl.setDeviceId(context, dcmConfig.getDeviceId());
+        isInitialized = true;
+    }
+
+    /**
+     * @return Whether or not the DCM Utilities have been properly initialized
+     */
+    public static boolean isInitialized() {
+        return isInitialized;
+    }
+
+    /**
+     * Disables the DCM library from usage and marks it as no longer initialized
+     */
+    public static void disable() {
+        isInitialized = false;
+    }
+
+    /**
+     * Add a MetricEvent object to the METRIC_INDEX using the metricId as a key.  If a MetricEvent already exists using the metricId, nothing is done.
+     */
+    public DCMetricsUtils createMetric(String metricId) {
+        return createMetric(metricId, MetricEventType.getDefault());
+    }
+
+    public DCMetricsUtils createMetric(String metricId, MetricEventType eventType) {
+        Slog.v(TAG, "Create Metric: " + metricId);
+
+        if (!METRIC_INDEX.containsKey(metricId)) {
+            MetricEvent metricEvent = mFactory.createMetricEvent(serviceName, methodName, eventType);
+            metricEvent.addString(DCMetricsConfig.SESSION_ID_LABEL, dcmConfig.getSessionId());
+            METRIC_INDEX.put(metricId, metricEvent);
+        } else {
+            Slog.d(TAG, "MetricEvent with metricId: " + metricId + " already exists.");
+        }
+
+        return this;
+    }
+
+    /**
+     * Get the raw MetricEvent object from the metric index.  For advanced metric manipulation or testing purposes.
+     */
+    public MetricEvent getMetricEvent(String metricId) {
+        Slog.v(TAG, "Get Metric: " + metricId);
+
+        return METRIC_INDEX.get(metricId);
+    }
+
+    /**
+     * Send a metricEvent to the DCM backend.  The MetricEvent is then removed from the index.
+     *
+     * Optional parameter priority governs how quickly the MetricsTransport service will send it.
+     *     Priority.HIGH means the service will send it right away.
+     *     Priority.NORMAL means the service may wait until it has several metrics to send and bundle them all at once.
+     */
+    public DCMetricsUtils recordMetric(String metricId) {
+        return recordMetric(metricId, DCMetricsConfig.DEFAULT_PRIORITY);
+    }
+
+    public DCMetricsUtils recordMetric(String metricId, Priority priority) {
+        Slog.v(TAG, "Record Metric: " + metricId + " with Priority: " + priority.toString());
+
+        if (METRIC_INDEX.containsKey(metricId)) {
+            MetricEvent metricEvent = METRIC_INDEX.get(metricId);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **148 and 149**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The removes that can remove the key are at lines: **154** and **169**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).